### PR TITLE
Grpo nan fix

### PR DESCRIPTION
**Pull request by pluesclues**
This is to fix gradient accumulation sizes above 1.

### CODE REVIEW

File: unsloth_zoo/rl_replacements.py [hunk 1]
```diff
@@ -60,7 +60,66 @@ def chunked_selective_log_softmax(logits, index):
...
+        final_mask = shift_mask & non_padding_mask
+
+        return final_mask
+
+def left_pack_padding(tensor: torch.Tensor, pad_id: int) -> torch.Tensor:
+    """
+    Moves all padding tokens in each sequence of a batch to the right.
+    """
+    mask = (tensor != pad_id)
+    sorted_indices = torch.argsort(mask, dim=1, descending=True)
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py]:
```suggestion
    sorted_indices = torch.argsort(mask, dim=1, descending=True, stable=True)
```

Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: This must be stable since the order does matter for gather

Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: See https://colab.research.google.com/drive/1gCTQZ0zLyDPqx6Gj3DOTpT-cs-tKl27-?usp=sharing

File: unsloth_zoo/rl_replacements.py [hunk 2]
```diff
@@ -60,7 +60,66 @@ def chunked_selective_log_softmax(logits, index):
...
+
+        prompt_section = input_ids[:, :-logits_to_keep]
+
+        padding_mask = (prompt_section == pad_token_id)
+
+        pad_token_counts = padding_mask.sum(dim=1)
+
+        return pad_token_counts
+
+def create_completion_attention_mask(
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py:82]: Fix indentation

File: unsloth_zoo/rl_replacements.py [hunk 3]
```diff
@@ -60,7 +60,66 @@ def chunked_selective_log_softmax(logits, index):
     return all_per_token_logps
 pass
 
+def calculate_pad_tokens_in_prompt(
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py:63]: Fix indentation - use spaces as well

File: unsloth_zoo/rl_replacements.py [hunk 4]
```diff
@@ -60,7 +60,66 @@ def chunked_selective_log_softmax(logits, index):
...
+    ) -> torch.Tensor:
+        """
+        Given prompt tensor, it returns all the left padded tokens in that sequence. so [pad, pad, pad, cat] = 3 tokens 
+        """
+        if logits_to_keep >= input_ids.shape[1]:
+            raise ValueError("logits_to_keep must be smaller than the sequence length.")
+
+        prompt_section = input_ids[:, :-logits_to_keep]
+
+        padding_mask = (prompt_section == pad_token_id)
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py:76]: Can we somehow re-use these binary masks? `~mask` etc

**Review by danielhanchen — CHANGES_REQUESTED** [review]
Nice work

File: unsloth_zoo/rl_replacements.py (hunk 1 above)
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: @pluesclues Here

File: unsloth_zoo/rl_replacements.py [hunk 5]
```diff
@@ -60,7 +60,66 @@ def chunked_selective_log_softmax(logits, index):
...
+    final_mask = shift_mask & non_padding_mask
+
+    return final_mask
+
+def left_pack_padding(tensor: torch.Tensor, pad_id: int) -> torch.Tensor:
+    """
+    Moves all padding tokens in each sequence of a batch to the right.
+    """
+    mask = (tensor != pad_id)
+    sorted_indices = sorted_indices = torch.argsort(mask, dim=1, descending=True, stable=True)
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py]: Double sorted indices

File: unsloth_zoo/rl_replacements.py (hunk 4 above)
Comment (pluesclues) [reply unsloth_zoo/rl_replacements.py:76]: This mask really is only used to get the left pad token counts and nothing else from the initial tokenized prompt id tensors. But other than that, we will not really use them any other time.

File: unsloth_zoo/rl_replacements.py (hunk 2 above)
Comment (pluesclues) [reply unsloth_zoo/rl_replacements.py:82]: I think i resolved this?

File: unsloth_zoo/rl_replacements.py (hunk 3 above)
Comment (pluesclues) [reply unsloth_zoo/rl_replacements.py:63]: I resolved the indentations, I am not sure where you want the spaces apologies
